notebooks/02-ingestion/05-partitioning-data.ipynb

```diff
@@ -0,0 +1,717 @@
+    "\n",
+    "![](assets/segments-iata.png)\n",
+    "\n",
+    "## Query Performance\n",
+    "Now that the data has been loaded using a few different partitioning strategies, you can use a test query to determine which one is the best.\n",
+    "In order to control the experiment we will turn cacheing off for the results of the query and measure the same query with each table.\n",
+    "\n",
+    "### The Query\n",
+    "It is well known that flight traffic follows a weekly pattern as well as a seasonal pattern. Given that we only have a month, let's take a look at the maximum and average delay for a single airline by day of the week so that we can see the pattern and how it affects delays. We'll only consider the departure delay for now and only use delays greater than 2 minutes as \"delayed\".\n",
+    "\n",
+    "One caveat about measuring performance on a laptop. The results obtained here will not necessarily be the best when the data grows and is distributed on a multi-node cluster. The intent here is to provide a way of thinking about optimizing the data. In a real cluster, it is likely that the results will vary and a different partitioning and clustering strategy will provide better results.\n",
+    "\n",
+    "The following cell defines a function we can use to measure performance:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "77a71e6a-73a2-4bd7-9e8d-fd7cb966da65",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from datetime import datetime\n",
+    "from statistics import mean \n",
+    "\n",
+    "def measure_query( sql: str, iterations: int ):\n",
+    "    req = sql_client.sql_request(sql)\n",
+    "    req.add_context(\"populateCache\", \"false\")  # run without cacheing results to get a real sense of performance\n",
+    "    req.add_context(\"useCache\", \"false\")  # do not use cached results\n",
+    "    stats = []\n",
+    "    while (iterations>0):\n",
+    "      start = datetime.now()\n",
+    "      sql_client.sql(req)\n",
+    "      end = datetime.now()\n",
+    "      stats.append( (end - start).total_seconds() * 10**3 ) # add run time in milliseconds\n",
+    "      iterations -=1\n",
+    "    return f\"Results = avg:{mean(stats)} ms   min:{min(stats)} ms  max:{max(stats)} ms\""
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "de7c2156-99bf-42bf-858e-537209c20fab",
+   "metadata": {},
+   "source": [
+    "Run the test SQL to see that in the month of November in 2005, American Airlines had the highest percentage of delayed flights on Sundays(7) although the worst average delay occurs on Fridays(5):"
```

Comment:
Run the cell below to set up the SQL that will be used for testing.

(You could add a bullet to the learn more section for them to try different SQL statements)

Reply:
good idea...done